--- lib/curl_setup.h	2015-06-11 19:50:43.000000000 +0200
+++ lib/curl_setup-PAT.h	2015-09-04 01:07:58.162035100 +0200
@@ -249,13 +249,16 @@
 #  ifdef HAVE_WINSOCK2_H
 #    include <winsock2.h>
 #    ifdef HAVE_WS2TCPIP_H
-#       include <ws2tcpip.h>
+#      include <ws2tcpip.h>
 #    endif
 #  else
 #    ifdef HAVE_WINSOCK_H
 #      include <winsock.h>
 #    endif
 #  endif
+#  ifdef USE_WINDOWS_SSPI
+#    include <wincrypt.h>
+#  endif
 #  include <tchar.h>
 #  ifdef UNICODE
      typedef wchar_t *(*curl_wcsdup_callback)(const wchar_t *str);
