diff --git a/lib/Makefile.m32 b/lib/Makefile.m32
index 02b31106c6e..37887eed5dd 100644
--- a/lib/Makefile.m32
+++ b/lib/Makefile.m32
@@ -298,7 +298,12 @@ endif
 ifdef ZSTD
   INCLUDES += -I"$(ZSTD_PATH)/include"
   CFLAGS += -DHAVE_ZSTD
-  DLL_LIBS += -L"$(ZSTD_PATH)/lib" -lzstd
+  DLL_LIBS += -L"$(ZSTD_PATH)/lib"
+  ifdef ZSTD_LIBS
+    DLL_LIBS += $(ZSTD_LIBS)
+  else
+    DLL_LIBS += -lzstd
+  endif
 endif
 ifdef BROTLI
   INCLUDES += -I"$(BROTLI_PATH)/include"
diff --git a/src/Makefile.m32 b/src/Makefile.m32
index afb4fd547bd..74e7b93d7ae 100644
--- a/src/Makefile.m32
+++ b/src/Makefile.m32
@@ -312,7 +312,12 @@ endif
 ifdef ZSTD
   INCLUDES += -I"$(ZSTD_PATH)/include"
   CFLAGS += -DHAVE_ZSTD
-  curl_LDADD += -L"$(ZSTD_PATH)/lib" -lzstd
+  curl_LDADD += -L"$(ZSTD_PATH)/lib"
+  ifdef ZSTD_LIBS
+    curl_LDADD += $(ZSTD_LIBS)
+  else
+    curl_LDADD += -lzstd
+  endif
 endif
 ifdef BROTLI
   INCLUDES += -I"$(BROTLI_PATH)/include"
