diff --git a/CMakeLists.txt b/CMakeLists.txt
index c0cc0c57b8..aae23b6acd 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -778,6 +778,8 @@ if(CURL_USE_OPENSSL)
       list(APPEND CURL_LIBS "stdc++")
       list(APPEND CMAKE_REQUIRED_LIBRARIES "stdc++")
     endif()
+  elseif(WIN32 AND CMAKE_COMPILER_IS_GNUCC AND HAVE_LIBRESSL)
+    list(APPEND CURL_LIBS "ws2_32")  # Add a duplicate to make binutils ld resolve symbols
   endif()
 
   if(HAVE_BORINGSSL)
@@ -1506,6 +1508,8 @@ endif()
 #
 # CA handling
 #
+option(CURL_CA_NATIVE_BY_DEFAULT "Use native CA store by default in the curl tool" OFF)
+
 if(_curl_ca_bundle_supported)
   set(CURL_CA_BUNDLE "auto" CACHE
     STRING "Path to the CA bundle. Set 'none' to disable or 'auto' for auto-detection. Defaults to 'auto'.")
diff --git a/lib/curl_config.h.cmake b/lib/curl_config.h.cmake
index 0838ddccfb..6be0b40ece 100644
--- a/lib/curl_config.h.cmake
+++ b/lib/curl_config.h.cmake
@@ -34,6 +34,9 @@
 /* Default SSL backend */
 #cmakedefine CURL_DEFAULT_SSL_BACKEND "${CURL_DEFAULT_SSL_BACKEND}"
 
+/* Use native CA store by default in curl tool */
+#cmakedefine CURL_CA_NATIVE_BY_DEFAULT 1
+
 /* disables alt-svc */
 #cmakedefine CURL_DISABLE_ALTSVC 1
 
diff --git a/src/tool_cfgable.c b/src/tool_cfgable.c
index 975ef2a458..5d71f1bd08 100644
--- a/src/tool_cfgable.c
+++ b/src/tool_cfgable.c
@@ -50,6 +50,12 @@ struct OperationConfig *config_alloc(struct GlobalConfig *global)
   config->file_clobber_mode = CLOBBER_DEFAULT;
   config->upload_flags = CURLULFLAG_SEEN;
   curlx_dyn_init(&config->postdata, MAX_FILE2MEMORY);
+
+#ifdef CURL_CA_NATIVE_BY_DEFAULT
+  config->native_ca_store = TRUE;
+  config->proxy_native_ca_store = TRUE;
+#endif
+
   return config;
 }
 
diff --git a/wcurl b/wcurl
index 7b53b74..45857d6 100755
--- a/scripts/wcurl
+++ b/scripts/wcurl
@@ -85,7 +85,7 @@ _EOF_
 # Display an error message and bail out.
 error()
 {
-    printf "%s\n" "$*" > /dev/stderr
+    printf "%s\n" "$*" >&2
     exit 1
