diff -u -r a/crypto/aes/aes-mingw64-x86_64.S b/crypto/aes/aes-mingw64-x86_64.S
--- a/crypto/aes/aes-mingw64-x86_64.S
+++ b/crypto/aes/aes-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/aes/aesni-mingw64-x86_64.S b/crypto/aes/aesni-mingw64-x86_64.S
--- a/crypto/aes/aesni-mingw64-x86_64.S
+++ b/crypto/aes/aesni-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/aes/aesni-sha1-mingw64-x86_64.S b/crypto/aes/aesni-sha1-mingw64-x86_64.S
--- a/crypto/aes/aesni-sha1-mingw64-x86_64.S
+++ b/crypto/aes/aesni-sha1-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/aes/bsaes-mingw64-x86_64.S b/crypto/aes/bsaes-mingw64-x86_64.S
--- a/crypto/aes/bsaes-mingw64-x86_64.S
+++ b/crypto/aes/bsaes-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/aes/vpaes-mingw64-x86_64.S b/crypto/aes/vpaes-mingw64-x86_64.S
--- a/crypto/aes/vpaes-mingw64-x86_64.S
+++ b/crypto/aes/vpaes-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/camellia/cmll-mingw64-x86_64.S b/crypto/camellia/cmll-mingw64-x86_64.S
--- a/crypto/camellia/cmll-mingw64-x86_64.S
+++ b/crypto/camellia/cmll-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/cpuid-mingw64-x86_64.S b/crypto/cpuid-mingw64-x86_64.S
--- a/crypto/cpuid-mingw64-x86_64.S
+++ b/crypto/cpuid-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/md5/md5-mingw64-x86_64.S b/crypto/md5/md5-mingw64-x86_64.S
--- a/crypto/md5/md5-mingw64-x86_64.S
+++ b/crypto/md5/md5-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/modes/ghash-mingw64-x86_64.S b/crypto/modes/ghash-mingw64-x86_64.S
--- a/crypto/modes/ghash-mingw64-x86_64.S
+++ b/crypto/modes/ghash-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/rc4/rc4-md5-mingw64-x86_64.S b/crypto/rc4/rc4-md5-mingw64-x86_64.S
--- a/crypto/rc4/rc4-md5-mingw64-x86_64.S
+++ b/crypto/rc4/rc4-md5-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/rc4/rc4-mingw64-x86_64.S b/crypto/rc4/rc4-mingw64-x86_64.S
--- a/crypto/rc4/rc4-mingw64-x86_64.S
+++ b/crypto/rc4/rc4-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/sha/sha1-mingw64-x86_64.S b/crypto/sha/sha1-mingw64-x86_64.S
--- a/crypto/sha/sha1-mingw64-x86_64.S
+++ b/crypto/sha/sha1-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/sha/sha256-mingw64-x86_64.S b/crypto/sha/sha256-mingw64-x86_64.S
--- a/crypto/sha/sha256-mingw64-x86_64.S
+++ b/crypto/sha/sha256-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/sha/sha512-mingw64-x86_64.S b/crypto/sha/sha512-mingw64-x86_64.S
--- a/crypto/sha/sha512-mingw64-x86_64.S
+++ b/crypto/sha/sha512-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
diff -u -r a/crypto/whrlpool/wp-mingw64-x86_64.S b/crypto/whrlpool/wp-mingw64-x86_64.S
--- a/crypto/whrlpool/wp-mingw64-x86_64.S
+++ b/crypto/whrlpool/wp-mingw64-x86_64.S
@@ -1,4 +1,4 @@
-#if defined(__CET__)
+#if defined(__CET__) && !defined(__MINGW32__)
 #include <cet.h>
 #else
 #define _CET_ENDBR
