--- a/src/openssl.h	2021-05-14 21:36:06.000000000 +0000
+++ b/src/openssl.h	2022-05-19 19:02:58.903387065 +0000
@@ -57,8 +57,9 @@
 #include <openssl/pem.h>
 #include <openssl/rand.h>
 
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L && \
-    !defined(LIBRESSL_VERSION_NUMBER)
+#if (OPENSSL_VERSION_NUMBER >= 0x10100000L && \
+    !defined(LIBRESSL_VERSION_NUMBER)) || \
+    LIBRESSL_VERSION_NUMBER >= 0x3050000fL
 # define HAVE_OPAQUE_STRUCTS 1
 #endif
 
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 50c028c45..cb8fee121 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -213,7 +213,7 @@ set(SOURCES
   userauth.h
   version.c)
 
-if(WIN32)
+if(WIN32 AND BUILD_SHARED_LIBS)
   list(APPEND SOURCES ${PROJECT_SOURCE_DIR}/win32/libssh2.rc)
 endif()
 
diff --git a/Makefile.am b/Makefile.am
index 986441bd68..b0b58d1824 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,6 +1,9 @@
 AUTOMAKE_OPTIONS = foreign nostdinc
 
-SUBDIRS = src tests docs
+SUBDIRS = src docs
+if ENABLE_TESTS
+SUBDIRS += tests
+endif
 if BUILD_EXAMPLES
 SUBDIRS += example
 endif
diff --git a/configure.ac b/configure.ac
index ca3cbee164..6fb7cc8260 100644
--- a/configure.ac
+++ b/configure.ac
@@ -265,6 +265,18 @@ AC_HELP_STRING([--disable-hidden-symbols],[Leave all symbols with default visibi
        AC_MSG_RESULT(no)
 )
 
+# Build tests?
+AC_ARG_ENABLE([tests],
+  [AS_HELP_STRING([--disable-tests], [Disable tests @<:@default=enabled@:>@])],
+  [
+    if ! test "x${enable_tests}" = "xyes"; then
+      enable_tests="no"
+    fi
+  ],
+  [enable_tests="yes"])
+AM_CONDITIONAL([ENABLE_TESTS], [test "x$enable_tests" = xyes])
+
+
 # Build example applications?
 AC_MSG_CHECKING([whether to build example applications])
 AC_ARG_ENABLE([examples-build],
