--- win32/GNUmakefile	2015-03-12 11:23:18.000000000 +0100
+++ win32/GNUmakefile-PAT	2015-09-03 15:28:51.806006400 +0200
@@ -111,11 +111,12 @@
 
 # Global flags for all compilers
 CFLAGS	= $(LIBSSH2_CFLAG_EXTRAS) $(OPT) -D$(DB) -DLIBSSH2_WIN32 # -DHAVE_CONFIG_H
+LDFLAGS = $(LIBSSH2_LDFLAG_EXTRAS)
 
 ifeq ($(CC),mwcc)
 LD	= mwld
 RC	= mwwinrc
-LDFLAGS	= -nostdlib
+LDFLAGS	+= -nostdlib
 AR	= $(LD)
 ARFLAGS	= -nostdlib -library -o
 LIBEXT	= lib
@@ -130,7 +131,7 @@
 else
 LD	= $(CROSSPREFIX)gcc
 RC	= $(CROSSPREFIX)windres
-LDFLAGS	= -s -shared -Wl,--output-def,$(TARGET).def,--out-implib,$(TARGET)dll.a
+LDFLAGS	+= -s -shared -Wl,--output-def,$(TARGET).def,--out-implib,$(TARGET)dll.a
 AR	= $(CROSSPREFIX)ar
 ARFLAGS	= -cq
 LIBEXT	= a
@@ -142,10 +143,12 @@
 CFLAGS	+= -fno-strict-aliasing
 CFLAGS	+= -Wall # -pedantic
 ifeq ($(ARCH),w64)
-CFLAGS	+= -D_AMD64_
+CFLAGS	+= -m64 -D_AMD64_
+LDFLAGS += -m64
 RCFLAGS	+= -F pe-x86-64
 else
 CFLAGS	+= -m32
+LDFLAGS += -m32
 RCFLAGS	+= -F pe-i386
 endif
 endif
