From 84014a0a81f762dbf61fe4bb4b745e4b7cd6de12 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <404610+tatsuhiro-t@users.noreply.github.com>
Date: Tue, 27 May 2025 00:35:49 +0900
Subject: Revert "Tighten up :path validation"

Reverts ngtcp2/nghttp3#359

Fixes #362

Closes #363

diff --git a/genpathchartbl.py b/genpathchartbl.py
deleted file mode 100755
index 6c26f23..0000000
--- a/genpathchartbl.py
+++ /dev/null
@@ -1,32 +0,0 @@
-#!/usr/bin/env python3
-import sys
-
-def name(i):
-    if i < 0x21:
-        return \
-            ['NUL ', 'SOH ', 'STX ', 'ETX ', 'EOT ', 'ENQ ', 'ACK ', 'BEL ',
-             'BS  ', 'HT  ', 'LF  ', 'VT  ', 'FF  ', 'CR  ', 'SO  ', 'SI  ',
-             'DLE ', 'DC1 ', 'DC2 ', 'DC3 ', 'DC4 ', 'NAK ', 'SYN ', 'ETB ',
-             'CAN ', 'EM  ', 'SUB ', 'ESC ', 'FS  ', 'GS  ', 'RS  ', 'US  ',
-             'SPC '][i]
-    elif i == 0x7f:
-        return 'DEL '
-
-for i in range(256):
-    if (0x41 <= i and i <= 0x5a) or (0x61 <= i and i <= 0x7a) or \
-       (0x30 <= i and i <= 0x39) or \
-       chr(i) in ['-', '.', '_', '~',
-                  '!', '$', '&', "'", '(', ')',
-                  '*', '+', ',', ';', '=',
-                  '/', '?',
-                  '%',
-                  ':', '@']:
-        sys.stdout.write('1 /* {}    */, '.format(chr(i)))
-    elif 0x80 <= i:
-        sys.stdout.write('0 /* {} */, '.format(hex(i)))
-    elif i == 0x7f or i <= 0x20:
-        sys.stdout.write('0 /* {} */, '.format(name(i)))
-    else:
-        sys.stdout.write('0 /* {}    */, '.format(chr(i)))
-    if (i + 1)%4 == 0:
-        sys.stdout.write('\n')
diff --git a/lib/nghttp3_http.c b/lib/nghttp3_http.c
index b3b01c5..9a8f06e 100644
--- a/lib/nghttp3_http.c
+++ b/lib/nghttp3_http.c
@@ -373,62 +373,62 @@ static char VALID_PATH_CHARS[] = {
   0 /* DC4  */, 0 /* NAK  */, 0 /* SYN  */, 0 /* ETB  */,
   0 /* CAN  */, 0 /* EM   */, 0 /* SUB  */, 0 /* ESC  */,
   0 /* FS   */, 0 /* GS   */, 0 /* RS   */, 0 /* US   */,
-  0 /* SPC  */, 1 /* !    */, 0 /* "    */, 0 /* #    */,
+  0 /* SPC  */, 1 /* !    */, 1 /* "    */, 1 /* #    */,
   1 /* $    */, 1 /* %    */, 1 /* &    */, 1 /* '    */,
   1 /* (    */, 1 /* )    */, 1 /* *    */, 1 /* +    */,
   1 /* ,    */, 1 /* -    */, 1 /* .    */, 1 /* /    */,
   1 /* 0    */, 1 /* 1    */, 1 /* 2    */, 1 /* 3    */,
   1 /* 4    */, 1 /* 5    */, 1 /* 6    */, 1 /* 7    */,
   1 /* 8    */, 1 /* 9    */, 1 /* :    */, 1 /* ;    */,
-  0 /* <    */, 1 /* =    */, 0 /* >    */, 1 /* ?    */,
+  1 /* <    */, 1 /* =    */, 1 /* >    */, 1 /* ?    */,
   1 /* @    */, 1 /* A    */, 1 /* B    */, 1 /* C    */,
   1 /* D    */, 1 /* E    */, 1 /* F    */, 1 /* G    */,
   1 /* H    */, 1 /* I    */, 1 /* J    */, 1 /* K    */,
   1 /* L    */, 1 /* M    */, 1 /* N    */, 1 /* O    */,
   1 /* P    */, 1 /* Q    */, 1 /* R    */, 1 /* S    */,
   1 /* T    */, 1 /* U    */, 1 /* V    */, 1 /* W    */,
-  1 /* X    */, 1 /* Y    */, 1 /* Z    */, 0 /* [    */,
-  0 /* \    */, 0 /* ]    */, 0 /* ^    */, 1 /* _    */,
-  0 /* `    */, 1 /* a    */, 1 /* b    */, 1 /* c    */,
+  1 /* X    */, 1 /* Y    */, 1 /* Z    */, 1 /* [    */,
+  1 /* \    */, 1 /* ]    */, 1 /* ^    */, 1 /* _    */,
+  1 /* `    */, 1 /* a    */, 1 /* b    */, 1 /* c    */,
   1 /* d    */, 1 /* e    */, 1 /* f    */, 1 /* g    */,
   1 /* h    */, 1 /* i    */, 1 /* j    */, 1 /* k    */,
   1 /* l    */, 1 /* m    */, 1 /* n    */, 1 /* o    */,
   1 /* p    */, 1 /* q    */, 1 /* r    */, 1 /* s    */,
   1 /* t    */, 1 /* u    */, 1 /* v    */, 1 /* w    */,
-  1 /* x    */, 1 /* y    */, 1 /* z    */, 0 /* {    */,
-  0 /* |    */, 0 /* }    */, 1 /* ~    */, 0 /* DEL  */,
-  0 /* 0x80 */, 0 /* 0x81 */, 0 /* 0x82 */, 0 /* 0x83 */,
-  0 /* 0x84 */, 0 /* 0x85 */, 0 /* 0x86 */, 0 /* 0x87 */,
-  0 /* 0x88 */, 0 /* 0x89 */, 0 /* 0x8a */, 0 /* 0x8b */,
-  0 /* 0x8c */, 0 /* 0x8d */, 0 /* 0x8e */, 0 /* 0x8f */,
-  0 /* 0x90 */, 0 /* 0x91 */, 0 /* 0x92 */, 0 /* 0x93 */,
-  0 /* 0x94 */, 0 /* 0x95 */, 0 /* 0x96 */, 0 /* 0x97 */,
-  0 /* 0x98 */, 0 /* 0x99 */, 0 /* 0x9a */, 0 /* 0x9b */,
-  0 /* 0x9c */, 0 /* 0x9d */, 0 /* 0x9e */, 0 /* 0x9f */,
-  0 /* 0xa0 */, 0 /* 0xa1 */, 0 /* 0xa2 */, 0 /* 0xa3 */,
-  0 /* 0xa4 */, 0 /* 0xa5 */, 0 /* 0xa6 */, 0 /* 0xa7 */,
-  0 /* 0xa8 */, 0 /* 0xa9 */, 0 /* 0xaa */, 0 /* 0xab */,
-  0 /* 0xac */, 0 /* 0xad */, 0 /* 0xae */, 0 /* 0xaf */,
-  0 /* 0xb0 */, 0 /* 0xb1 */, 0 /* 0xb2 */, 0 /* 0xb3 */,
-  0 /* 0xb4 */, 0 /* 0xb5 */, 0 /* 0xb6 */, 0 /* 0xb7 */,
-  0 /* 0xb8 */, 0 /* 0xb9 */, 0 /* 0xba */, 0 /* 0xbb */,
-  0 /* 0xbc */, 0 /* 0xbd */, 0 /* 0xbe */, 0 /* 0xbf */,
-  0 /* 0xc0 */, 0 /* 0xc1 */, 0 /* 0xc2 */, 0 /* 0xc3 */,
-  0 /* 0xc4 */, 0 /* 0xc5 */, 0 /* 0xc6 */, 0 /* 0xc7 */,
-  0 /* 0xc8 */, 0 /* 0xc9 */, 0 /* 0xca */, 0 /* 0xcb */,
-  0 /* 0xcc */, 0 /* 0xcd */, 0 /* 0xce */, 0 /* 0xcf */,
-  0 /* 0xd0 */, 0 /* 0xd1 */, 0 /* 0xd2 */, 0 /* 0xd3 */,
-  0 /* 0xd4 */, 0 /* 0xd5 */, 0 /* 0xd6 */, 0 /* 0xd7 */,
-  0 /* 0xd8 */, 0 /* 0xd9 */, 0 /* 0xda */, 0 /* 0xdb */,
-  0 /* 0xdc */, 0 /* 0xdd */, 0 /* 0xde */, 0 /* 0xdf */,
-  0 /* 0xe0 */, 0 /* 0xe1 */, 0 /* 0xe2 */, 0 /* 0xe3 */,
-  0 /* 0xe4 */, 0 /* 0xe5 */, 0 /* 0xe6 */, 0 /* 0xe7 */,
-  0 /* 0xe8 */, 0 /* 0xe9 */, 0 /* 0xea */, 0 /* 0xeb */,
-  0 /* 0xec */, 0 /* 0xed */, 0 /* 0xee */, 0 /* 0xef */,
-  0 /* 0xf0 */, 0 /* 0xf1 */, 0 /* 0xf2 */, 0 /* 0xf3 */,
-  0 /* 0xf4 */, 0 /* 0xf5 */, 0 /* 0xf6 */, 0 /* 0xf7 */,
-  0 /* 0xf8 */, 0 /* 0xf9 */, 0 /* 0xfa */, 0 /* 0xfb */,
-  0 /* 0xfc */, 0 /* 0xfd */, 0 /* 0xfe */, 0 /* 0xff */
+  1 /* x    */, 1 /* y    */, 1 /* z    */, 1 /* {    */,
+  1 /* |    */, 1 /* }    */, 1 /* ~    */, 0 /* DEL  */,
+  1 /* 0x80 */, 1 /* 0x81 */, 1 /* 0x82 */, 1 /* 0x83 */,
+  1 /* 0x84 */, 1 /* 0x85 */, 1 /* 0x86 */, 1 /* 0x87 */,
+  1 /* 0x88 */, 1 /* 0x89 */, 1 /* 0x8a */, 1 /* 0x8b */,
+  1 /* 0x8c */, 1 /* 0x8d */, 1 /* 0x8e */, 1 /* 0x8f */,
+  1 /* 0x90 */, 1 /* 0x91 */, 1 /* 0x92 */, 1 /* 0x93 */,
+  1 /* 0x94 */, 1 /* 0x95 */, 1 /* 0x96 */, 1 /* 0x97 */,
+  1 /* 0x98 */, 1 /* 0x99 */, 1 /* 0x9a */, 1 /* 0x9b */,
+  1 /* 0x9c */, 1 /* 0x9d */, 1 /* 0x9e */, 1 /* 0x9f */,
+  1 /* 0xa0 */, 1 /* 0xa1 */, 1 /* 0xa2 */, 1 /* 0xa3 */,
+  1 /* 0xa4 */, 1 /* 0xa5 */, 1 /* 0xa6 */, 1 /* 0xa7 */,
+  1 /* 0xa8 */, 1 /* 0xa9 */, 1 /* 0xaa */, 1 /* 0xab */,
+  1 /* 0xac */, 1 /* 0xad */, 1 /* 0xae */, 1 /* 0xaf */,
+  1 /* 0xb0 */, 1 /* 0xb1 */, 1 /* 0xb2 */, 1 /* 0xb3 */,
+  1 /* 0xb4 */, 1 /* 0xb5 */, 1 /* 0xb6 */, 1 /* 0xb7 */,
+  1 /* 0xb8 */, 1 /* 0xb9 */, 1 /* 0xba */, 1 /* 0xbb */,
+  1 /* 0xbc */, 1 /* 0xbd */, 1 /* 0xbe */, 1 /* 0xbf */,
+  1 /* 0xc0 */, 1 /* 0xc1 */, 1 /* 0xc2 */, 1 /* 0xc3 */,
+  1 /* 0xc4 */, 1 /* 0xc5 */, 1 /* 0xc6 */, 1 /* 0xc7 */,
+  1 /* 0xc8 */, 1 /* 0xc9 */, 1 /* 0xca */, 1 /* 0xcb */,
+  1 /* 0xcc */, 1 /* 0xcd */, 1 /* 0xce */, 1 /* 0xcf */,
+  1 /* 0xd0 */, 1 /* 0xd1 */, 1 /* 0xd2 */, 1 /* 0xd3 */,
+  1 /* 0xd4 */, 1 /* 0xd5 */, 1 /* 0xd6 */, 1 /* 0xd7 */,
+  1 /* 0xd8 */, 1 /* 0xd9 */, 1 /* 0xda */, 1 /* 0xdb */,
+  1 /* 0xdc */, 1 /* 0xdd */, 1 /* 0xde */, 1 /* 0xdf */,
+  1 /* 0xe0 */, 1 /* 0xe1 */, 1 /* 0xe2 */, 1 /* 0xe3 */,
+  1 /* 0xe4 */, 1 /* 0xe5 */, 1 /* 0xe6 */, 1 /* 0xe7 */,
+  1 /* 0xe8 */, 1 /* 0xe9 */, 1 /* 0xea */, 1 /* 0xeb */,
+  1 /* 0xec */, 1 /* 0xed */, 1 /* 0xee */, 1 /* 0xef */,
+  1 /* 0xf0 */, 1 /* 0xf1 */, 1 /* 0xf2 */, 1 /* 0xf3 */,
+  1 /* 0xf4 */, 1 /* 0xf5 */, 1 /* 0xf6 */, 1 /* 0xf7 */,
+  1 /* 0xf8 */, 1 /* 0xf9 */, 1 /* 0xfa */, 1 /* 0xfb */,
+  1 /* 0xfc */, 1 /* 0xfd */, 1 /* 0xfe */, 1 /* 0xff */
 };
 
 static int check_path(const uint8_t *value, size_t len) {
diff --git a/tests/nghttp3_conn_test.c b/tests/nghttp3_conn_test.c
index 35b8384..932251a 100644
--- a/tests/nghttp3_conn_test.c
+++ b/tests/nghttp3_conn_test.c
@@ -2097,14 +2097,6 @@ void test_nghttp3_conn_http_req_header(void) {
     MAKE_NV(":method", "GET"),
     MAKE_NV(":authority", "localhost"),
   };
-  /* :path contains all allowed characters. */
-  const nghttp3_nv allowedcharpath_reqnv[] = {
-    MAKE_NV(":scheme", "https"),
-    MAKE_NV(":path", "/ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123"
-                     "456789-._~!$&'()*+,;=?%:@"),
-    MAKE_NV(":method", "GET"),
-    MAKE_NV(":authority", "localhost"),
-  };
   /* HEAD method is used. */
   const nghttp3_nv headmethod_reqnv[] = {
     MAKE_NV(":scheme", "https"),
@@ -2243,8 +2235,6 @@ void test_nghttp3_conn_http_req_header(void) {
   check_http_req_header(emptypath_reqnv, nghttp3_arraylen(emptypath_reqnv), 0);
   check_http_req_header(badcharpath_reqnv, nghttp3_arraylen(badcharpath_reqnv),
                         NGHTTP3_ERR_MALFORMED_HTTP_HEADER);
-  check_http_req_header(allowedcharpath_reqnv,
-                        nghttp3_arraylen(allowedcharpath_reqnv), 0);
   check_http_req_header(headmethod_reqnv, nghttp3_arraylen(headmethod_reqnv),
                         0);
   check_http_req_header(dupproto_reqnv, nghttp3_arraylen(dupproto_reqnv),
