diff --git a/CMakeLists.txt b/CMakeLists.txt
index c5a3d9c1..06ab1ed7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -134,7 +134,7 @@ if(HAVE_CUNIT)
   add_custom_target(check COMMAND ${CMAKE_CTEST_COMMAND})
 endif()
 
-# OpenSSL (required for libngtcp2_crypto_openssl,
+# OpenSSL (required for libngtcp2_crypto_quictls,
 # libngtcp2_crypto_picotls and examples)
 include(CheckSymbolExists)
 if(ENABLE_OPENSSL AND OPENSSL_FOUND)
diff --git a/configure.ac b/configure.ac
index cf29215c..227e551c 100644
--- a/configure.ac
+++ b/configure.ac
@@ -227,7 +227,7 @@ AM_CONDITIONAL([HAVE_CUNIT], [ test "x${have_cunit}" = "xyes" ])
 
 AM_CONDITIONAL([ENABLE_SHARED], [ test "x${enable_shared}" = "xyes" ])
 
-# OpenSSL (required for libngtcp2_crypto_openssl,
+# OpenSSL (required for libngtcp2_crypto_quictls,
 # libngtcp2_crypto_picotls and examples)
 have_openssl=no
 have_vanilla_openssl=no
diff --git a/crypto/includes/CMakeLists.txt b/crypto/includes/CMakeLists.txt
index 10f91222..477c10d1 100644
--- a/crypto/includes/CMakeLists.txt
+++ b/crypto/includes/CMakeLists.txt
@@ -27,7 +27,7 @@ install(FILES
 
 if(HAVE_OPENSSL)
   install(FILES
-      ngtcp2/ngtcp2_crypto_openssl.h
+      ngtcp2/ngtcp2_crypto_quictls.h
       DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/ngtcp2")
 endif()
 
diff --git a/crypto/quictls/.gitignore b/crypto/quictls/.gitignore
index 296fddc1..6f4bcf87 100644
--- a/crypto/quictls/.gitignore
+++ b/crypto/quictls/.gitignore
@@ -1 +1 @@
-/libngtcp2_crypto_openssl.pc
+/libngtcp2_crypto_quictls.pc
diff --git a/crypto/shared.h b/crypto/shared.h
index 8c7c88d0..d69fd212 100644
--- a/crypto/shared.h
+++ b/crypto/shared.h
@@ -100,7 +100,7 @@ ngtcp2_crypto_ctx *ngtcp2_crypto_ctx_initial(ngtcp2_crypto_ctx *ctx);
  * `ngtcp2_crypto_aead_init` initializes |aead| with the provided
  * |aead_native_handle| which is an underlying AEAD object.
  *
- * If libngtcp2_crypto_openssl is linked, |aead_native_handle| must be
+ * If libngtcp2_crypto_quictls is linked, |aead_native_handle| must be
  * a pointer to EVP_CIPHER.
  *
  * If libngtcp2_crypto_gnutls is linked, |aead_native_handle| must be
@@ -214,7 +214,7 @@ int ngtcp2_crypto_update_traffic_secret(uint8_t *dest, uint32_t version,
  * pointed by |buf| of length |len|, to the native handle |tls|.
  *
  * |tls| points to a implementation dependent TLS session object.  If
- * libngtcp2_crypto_openssl is linked, |tls| must be a pointer to SSL
+ * libngtcp2_crypto_quictls is linked, |tls| must be a pointer to SSL
  * object.
  *
  * This function returns 0 if it succeeds, or -1.
@@ -230,7 +230,7 @@ int ngtcp2_crypto_set_local_transport_params(void *tls, const uint8_t *buf,
  * `ngtcp2_conn_set_remote_transport_params`.
  *
  * |tls| points to a implementation dependent TLS session object.  If
- * libngtcp2_crypto_openssl is linked, |tls| must be a pointer to SSL
+ * libngtcp2_crypto_quictls is linked, |tls| must be a pointer to SSL
  * object.
  *
  * This function returns 0 if it succeeds, or -1.
