--- a/Configurations/10-main.conf	2016-02-13 14:54:09.000000000 +0100
+++ b/Configurations/10-main.conf-PAT	2016-02-13 14:54:51.000000000 +0100
@@ -1174,7 +1174,7 @@
         perlasm_scheme   => "coff",
         dso_scheme       => "win32",
         shared_target    => "cygwin-shared",
-        shared_cflag     => "-D_WINDLL -DOPENSSL_USE_APPLINK",
+        shared_cflag     => "-D_WINDLL",
         shared_ldflag    => "-static-libgcc",
         shared_rcflag    => "--target=pe-i386",
         shared_extension => ".dll",
